### PR TITLE
Custom JavaScript string rust type

### LINKED ISSUE #1377: Custom JavaScript string rust type
The current JavaScript string rust (`RcString` which holds a `Rc<str>`) is 16 bytes this is because it stores the pointer to the data and length on the stack. And this prevents #1373 from being implemented. So we need to write a custom string type which is referenced counted like `Rc<str>` and holds the length on the heap.

I should almost be done with this, and I'll create a PR soon :)

### PR DESCRIPTION
**Pull request by HalidOdat**
This Pull Request fixes/closes #1377 .

As per #1376 `RcString` has been renamed to `JsString`.

This is similar to `Rc<str>`. But unlike `Rc<str>` which stores the length on the stack and a pointer to the data The `JsString` length and data is stored on the heap. (`Rc<str>` keeps refcount on the heap anyway) and just an non-null pointer is kept on the stack, so its size is the size of a pointer.

Also added `JsString::concat(x, y)` function, so we don't have to do `JsString::new(format!("{}{}", x, y))` which wastes an allocation.

Moved `src/value/rcstring.rs` **=>** `src/string.rs` (like `symbol.rs`)
Reexport `JsString` to root level (because its such a common import, like `Value`)

This should unblock #1373 :)

### CODE REVIEW

File: boa/src/value/rcstring.rs [hunk 1]
```diff
@@ -1,114 +1,448 @@
...
 
-unsafe impl Trace for RcString {
-    empty_trace!();
+    /// The number of references to the string.
+    ///
+    /// When this reaches `0` the string is deallocated.
+    refcount: Cell<usize>,
+
+    /// An empty array which is used to get the offset of string data.
+    data: [u8; 0],
```
Comment (Razican) [inline boa/src/value/rcstring.rs:24]: Could we use something like [`NonNull`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html) for this?

Comment (HalidOdat) [reply boa/src/value/rcstring.rs:24]: But we aren't pointing to any other allocation, it's only one allocation for len, refcount and string data, and size of `NonNull` is size of pointer while size of `[u8; 0]` is zero, we just use it to get the offset of the string bytes. by calling `data.as_ptr()`. This pattern is also used in C/C++ [Flexible array member](https://en.wikipedia.org/wiki/Flexible_array_member)

**Comment by HalidOdat** [comment]
There seems to be an overall nice increase in performance :)

**Comment by neeldug** [comment]
@HalidOdat just wondering why there should be this strong of a performance improvement, as I'd expect the memory usage to decrease, but due to heap allocations and deallocations being slower than stack shouldn't it have the adverse effect on the runtime?

**Comment by HalidOdat** [comment]
> @HalidOdat just wondering why there should be this strong of a performance improvement, as I'd expect the memory usage to decrease, but due to heap allocations and deallocations being slower than stack shouldn't it have the adverse effect on the runtime?

`JsString::concat` saves one allocation so that's probably why string concatenation improved Besides that we don't really allocate/deallocate anymore than we did previously. In the benchmarks we can see that the main improvement is in arithmetic operations this is most likely do to the smaller size of `Value` (from `24` to `16`), there are probably other places that have smaller size now, like `PropertyKey`, `Symbol`, string property entries, etc.

**Pull request merged**